### ISSUE TITLE
TypeGuard is considered a subtype of bool

### ISSUE THREAD

**Issue by DaviDevMod** [microsoft/pyright#4902]
**Describe the bug**
According to [the docs](https://peps.python.org/pep-0647/#typeguard-type), an error should be raised by type checkers when trying to assign a `Callable[..., TypeGuard]` to a `Callable[..., bool]`:

> In all other respects, TypeGuard is a distinct type from bool. It is not a subtype of bool. Therefore, `Callable[..., TypeGuard[int]]` is not assignable to `Callable[..., bool]`.

But pyright does allow the assignment.

**Expected behavior**
Pyright should raise an error when I try to assign a `Callable[..., TypeGuard]` to a `Callable[..., bool]`

**Screenshots or Code**
[mypy](https://mypy-play.net/?mypy=latest&python=3.11&gist=3e0a52aa16e6df0451cad99398638b09) behaves like pyright, on this issue.

```python
from typing import Literal, TypeGuard, get_args
from collections.abc import Callable


Operation = Literal["+", "-", "*", "/"]


def is_coercible_to_float(value: str):
    try:
        float(value)
        return True
    except ValueError:
        return False


def is_operation(value: str) -> TypeGuard[Operation]:
    return value in get_args(Operation)


def f(value, is_right_type: Callable[..., bool]):
    return "YES" if is_right_type(value) else "NO"


# OK cause `is_coercible_to_float` returns a `bool`.
f("0.5sss", is_coercible_to_float)

# Should error cause `is_operation` return a `TypeGuard`
# which is not a subtype of `bool`.
f("+", is_operation)

# Again, an error should be raised.
a: Callable[..., bool] = is_operation

# These are not raising an error either.
# But I guess it's the correct behaviour
# (cause `TypeGuard` is probably a subtype of `int`).
b: Callable[..., int] = is_operation
c: Callable[..., float] = is_operation
```

**VS Code extension or command-line**
I am using pyright via pylance v2023.3.40 as a VS Code extension.

**Additional context**
The current behaviour is more convenient than that outlined by the docs, since it allows a `Callable[..., bool]` signature to do the job of an unnecessarily verbose `Callable[..., bool | TypeGuard[Operation]]` signature.
So personally, I am not really looking for a fix, but the bug deserves to be documented somewhere.

**Comment by hmc-cs-mdrissi** [comment]
There's a thread on this topic in typing-sig [here](https://mail.python.org/archives/list/typing-sig@python.org/thread/ZDF2GS54VBHBNMVKEYWDHMGIYTN4XOHZ/). It's not really resolved though and I think current behavior makes more sense.

**Comment by erictraut** [comment]
Yeah, I think pyright's behavior is correct here. I'm the author of PEP 647, and pyright is the reference implementation for the PEP. I consider the text in the PEP to be incorrect on this particular point. We should probably go back and amend the PEP to clarify this point.

**Comment by DaviDevMod** [comment]
@hmc-cs-mdrissi It's nice to know about typing-sig, thanks.

And yes, the current behaviour definitely makes more sense.

@erictraut glad that the bug is actually in the docs. If you need manpower for the fix I am available.
